File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/HomeFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.BibleClient
import com.nexussoft.verbum.clients.NotificationClient
import com.nexussoft.verbum.clients.PreferencesClient
import com.nexussoft.verbum.common.arch.Effect
import com.nexussoft.verbum.common.arch.Reducer
import com.nexussoft.verbum.common.arch.combine
import com.nexussoft.verbum.common.arch.only
import com.nexussoft.verbum.common.arch.pullback
import com.nexussoft.verbum.common.arch.with
import com.nexussoft.verbum.models.PassageReference
import java.time.Clock
import java.time.LocalDateTime

/** Home (§5): a question, a way back into what you were reading, the verse of the day. Twin of iOS `HomeFeature`. */
object HomeFeature {
    data class State(
        val lastRead: PassageReference? = null,
        val greeting: Greeting = Greeting.MORNING,
        val dailyVerse: DailyVerseFeature.State = DailyVerseFeature.State(),
    )

    enum class Greeting {
        MORNING, AFTERNOON, EVENING;

        companion object {
            fun at(hour: Int) = when (hour) {
                in 5..11 -> MORNING
                in 12..17 -> AFTERNOON
                else -> EVENING
            }
        }
    }

    sealed interface Action {
        data object Started : Action
        data object ContinueReadingTapped : Action
        data object SearchTapped : Action
        data object ArrivalTapped : Action
        data class DailyVerse(val action: DailyVerseFeature.Action) : Action
        data class Delegate(val delegate: DelegateAction) : Action
    }

    sealed interface DelegateAction {
        data class OpenPassage(val reference: PassageReference) : DelegateAction
        data object OpenSearch : DelegateAction
        data object OpenArrival : DelegateAction
    }

    fun reducer(
        preferences: PreferencesClient,
        clock: Clock = Clock.systemDefaultZone(),
        bibleClient: BibleClient,
        notifications: NotificationClient,
        dailyVerseTitle: () -> String = { "Verse of the day" },
    ): Reducer<State, Action> = combine(
        DailyVerseFeature.reducer(bibleClient, notifications, preferences, clock, dailyVerseTitle).pullback(
            get = { it.dailyVerse }, set = { s, c -> s.copy(dailyVerse = c) },
            extractAction = { (it as? Action.DailyVerse)?.action }, embedAction = { Action.DailyVerse(it) },
        ),
        Reducer { state, action ->
            when (action) {
                Action.Started -> {
                    val now = LocalDateTime.now(clock)
                    state.copy(
                        lastRead = LastRead.decode(preferences.string(ChapterReaderFeature.LAST_READ_KEY)),
                        greeting = Greeting.at(now.hour),
                    ).only()
                }
                Action.ContinueReadingTapped -> state.lastRead?.let { state.with(Effect.Send(Action.Delegate(DelegateAction.OpenPassage(it)))) } ?: state.only()
                Action.SearchTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenSearch)))
                Action.ArrivalTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenArrival)))
                is Action.DailyVerse -> when (val d = (action.action as? DailyVerseFeature.Action.Delegate)?.delegate) {
                    is DailyVerseFeature.DelegateAction.OpenPassage -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenPassage(d.reference))))
                    null -> state.only()
                }
                is Action.Delegate -> state.only()
            }
        },
    )
}

File: VerbumKit/Sources/Features/Home/HomeFeature.swift
import ComposableArchitecture
import Foundation
import Models

/// Home (spec §5): a question, a way back into what you were reading, and the
/// verse of the day. Deliberately small (§73.8).
@Reducer
public struct HomeFeature {
    @ObservableState
    public struct State: Equatable {
        @Shared(.lastRead) public var lastRead
        public var greeting: Greeting = .morning
        public var dailyVerse = DailyVerseFeature.State()

        public init() {}
    }

    public enum Greeting: Equatable, Sendable {
        case morning, afternoon, evening

        static func at(hour: Int) -> Greeting {
            switch hour {
            case 5..<12: .morning
            case 12..<18: .afternoon
            default: .evening
            }
        }
    }

    public enum Action: Equatable {
        case task
        case continueReadingTapped
        case searchTapped
        case arrivalTapped
        case dailyVerse(DailyVerseFeature.Action)
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case openPassage(PassageReference)
            case openSearch
            case openArrival
        }
    }

    @Dependency(\.date.now) var now
    @Dependency(\.calendar) var calendar

    public init() {}

    public var body: some ReducerOf<Self> {
        Scope(state: \.dailyVerse, action: \.dailyVerse) { DailyVerseFeature() }
        Reduce { state, action in
            switch action {
            case .task:
                state.greeting = Greeting.at(hour: calendar.component(.hour, from: now))
                return .none

            case .continueReadingTapped:
                guard let reference = state.lastRead else { return .none }
                return .send(.delegate(.openPassage(reference)))

            case .dailyVerse(.delegate(.openPassage(let reference))):
                return .send(.delegate(.openPassage(reference)))

            case .dailyVerse:
                return .none

            case .searchTapped:
                return .send(.delegate(.openSearch))
            case .arrivalTapped:
                return .send(.delegate(.openArrival))

            case .delegate:
                return .none
            }
        }
    }
}
